design-docs/opmodes.md

```diff
@@ -0,0 +1,516 @@
+  public static int removeOpModeOption(String name) {...}
+  public static void clearOpModeOptions() {...}
+}
+```
+
+## Java Robot Code Examples
+
+Non-command-based robots will typically use classes to define opmodes.
+
+The following example code for non-command-based Java demonstrates the following:
+- Robot class
+- A periodic autonomous opmode
+- A periodic teleop opmode
+- A linear test opmode
+
+Robot:
+
+```java
+public class Robot extends RobotBase {
+  public final DifferentialDrive drive = new DifferentialDrive(...);
+
+  public Robot() {}
+}
+```
+
+Autonomous opmode:
+
+```java
+@Autonomous(name="Drive straight", group="Drive")
+public class AutoDriveStraight extends PeriodicOpMode {
+  private final Robot robot;
+  private final Timer timer = new Timer();
+
+  public AutoDriveStraight(Robot robot) {
+    this.robot = robot;
+  }
+
+  @Override
+  public void start() {
+    timer.start();
+  }
+
+  @Override
+  public void periodic() {
+    // drive forward for 2 seconds at half speed, then stop
+    if (!timer.hasElapsed(2.0)) {
+      robot.drive.arcadeDrive(0.5, 0);
+    } else {
+      robot.drive.arcadeDrive(0, 0);
+    }
+  }
+}
+```
+
+Teleop opmode:
+
+```java
+@Teleoperated
+public class Teleop extends PeriodicOpMode {
+  private final Joystick joy = new Joystick(1);
+  private final Robot robot;
+
+  public Teleop(Robot robot) {
+    this.robot = robot;
+  }
+
+  @Override
+  public void periodic() {
+    robot.drive.arcadeDrive(joy.getY(), joy.getX());
+  }
+}
+```
+
+Test opmode:
+
+```java
+@TestOpMode("Blink dashboard indicator")
+public class TestDashboardIndicator extends LinearOpMode {
+  @Override
+  public void run() {
+    Telemetry.log("indicator", true);
+    Timer.sleep(0.5);
+    Telemetry.log("indicator", false);
+  }
+}
+```
+
+## C++ Robot Code
+
+C++ does not support annotations, so explicit registration via `RobotBase` function calls is required.
+
+## Python Robot Code
+
+Should be able to be annotation (decorator) based, similar to Java.
+
+## HAL
+
+At the HAL level, Control Word bits indicate enabled state and teleop/auto/test selection (identical to current WPILib).  What needs to be added are functions to maintain the list of available opmodes and get the selected/active opmodes as communicated by the DS.
+
+Adding the following functions provides all the required functionality:
+- `int AddOpModeOption(string name, string group, string description, int flags)` – returns a unique ID identifying the opmode (1:1 mapping to name).  Flags indicates which list the opmode appears on.
+- `int RemoveOpModeOption(string name)` – returns the unique ID previously associated to that opmode name
+- `void ClearOpModeOptions()` – clears all opmodes
+- `int GetActiveOpMode()` - gets the currently active opmode ID; returns 0 when the robot is disabled
+- `int GetSelectedAutonomousOpMode()` - gets the opmode ID currently selected for autonomous; valid when the robot is disabled; returns 0 if unknown or no selection
```

Comment:
What does this return if you have  separate auto op-modes selected for just auto vs the auto portion of match?

Reply:
The protocol only has one location to hold what this function returns.  So this is really more of DS behavior question when you change from auto to match.